### PR TITLE
Fix linalg.norm sample inputs function and related failures

### PR DESCRIPTION
**Pull request by peterbell10**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #84631
* #84455
* #84567
* #84454
* #84554
* #84696
* #84453
* __->__ #84452

Due to an indentation error, the return statement happens after just 1
loop of `for test_size in test_sizes` so only one shape was ever
tested.

This also revealed several cases where the provided shapes don't work
so I've disabled the generation of those sample inputs.

### CODE REVIEW

File: torch/testing/_internal/opinfo/definitions/linalg.py [hunk 1]
```diff
@@ -555,6 +558,28 @@ def sample_inputs_linalg_norm(
...
+                        2: (0, 1),
+                        1: (1,),
+                        -1: (1,),
+                        -2: (0, 1),
+                        -np.inf: (0,),
+                    }.get(ord, ())
+
+                    if any(test_size[d] == 0 for d in dims_to_check):
+                        # IndexError: amax(): Expected reduction dim {dim} to
+                        # have non-zero size.
```
Comment (Lezcano) [inline torch/testing/_internal/opinfo/definitions/linalg.py:580]: This doesn't happen in the `"fro"` and `"nuc"` cases.

File: torch/testing/_internal/opinfo/definitions/linalg.py [hunk 2]
```diff
@@ -555,6 +558,28 @@ def sample_inputs_linalg_norm(
...
+                elif is_matrix_norm:
+                    dims_to_check = {
+                        None: (0,),
+                        np.inf: (0,),
+                        2: (0, 1),
+                        1: (1,),
+                        -1: (1,),
+                        -2: (0, 1),
+                        -np.inf: (0,),
+                    }.get(ord, ())
```
Comment (Lezcano) [inline torch/testing/_internal/opinfo/definitions/linalg.py:576]: Are `nuc` and `fro` not considered here? And if so, why?

**Review by lezcano — COMMENTED** [review]
Uf, good catch. 
That function could do with some refactoring by using `make_arg` and turning it into a generator for clarity.

I left a couple comments, but otherwise it looks good. I can put up a follow-up PR fixing all these xfails.

**Comment by Lezcano** [comment]
fwiw, for a good high-level summary of what this function does, see the reference implementation at https://github.com/pytorch/pytorch/blob/master/torch/_refs/linalg/__init__.py

File: torch/testing/_internal/opinfo/definitions/linalg.py (hunk 2 above)
Comment (peterbell10) [reply torch/testing/_internal/opinfo/definitions/linalg.py:576]: The `fro` and `nuc` cases work for all sizes, so we go to the default `()` i.e. not checking any dims. I could spell it out explicitly if you prefer.

Comment (Lezcano) [reply torch/testing/_internal/opinfo/definitions/linalg.py:576]: Ah, right. For some reason I thought we were skipping that case.

**Review by lezcano — APPROVED** [review]
Cool, thanks!

**Comment by Lezcano** [comment]
Could we have an OK on this one @zou3519? My linalg powers are not enough to get this one merged as it touches functorch.

File: functorch/test/test_vmap.py [hunk 3]
```diff
@@ -3304,6 +3304,8 @@ def test_vmap_exhaustive(self, device, dtype, op):
         xfail('isin'),
         xfail('linalg.matrix_rank'),
         xfail('linalg.matrix_rank', 'hermitian'),
+        xfail('linalg.norm'),
+        xfail('linalg.norm', 'subgradients_at_zero'),
```
[the remark this answers is not in the document]
Comment (peterbell10) [reply functorch/test/test_vmap.py]: @zou3519 PTAL, I've applied your fix.

**Review by zou3519 — APPROVED** [review]

File: functorch/test/test_vmap.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (zou3519) [reply functorch/test/test_vmap.py]: LGTM, thanks for applying the fix

**Pull request closed without merging**